File: Back-laravel/app/Services/AuthService.php
<?php

namespace App\Services;

use App\Helpers\CryptoHelper;
use App\Models\EmailVerificationToken;
use App\Models\LoginAttempt;
use App\Models\PasswordResetToken;
use App\Models\RefreshToken;
use App\Models\User;
use Illuminate\Support\Facades\Cache;
use Illuminate\Support\Facades\Hash;
use Illuminate\Support\Facades\Log;

class AuthService
{
    public function __construct(
        private TokenService $tokenService,
        private EmailService $emailService
    ) {}

    // ─── Register ──────────────────────────────────────────────────────────────
    public function register(array $data): array
    {
        // 1. Uniqueness checks handled by FormRequest

        // 2. Hash password with Argon2id
        $passwordHash = Hash::driver('argon2id')->make($data['password']);

        // 3. Create user
        $user = User::create([
            'id'            => CryptoHelper::generateCuid(),
            'first_name'    => $data['firstName'],
            'last_name'     => $data['lastName'],
            'email'         => $data['email'],
            'phone'         => $data['phone'] ?? null,
            'password_hash' => $passwordHash,
            'role'          => $data['role'] ?? 'CANDIDATE',
            'status'        => 'PENDING_VERIFICATION',
        ]);

        // 4. Generate email verification token
        $rawToken = CryptoHelper::generateSecureToken(32);
        $tokenHash = CryptoHelper::hashToken($rawToken);

        EmailVerificationToken::create([
            'id'         => CryptoHelper::generateCuid(),
            'user_id'    => $user->id,
            'token_hash' => $tokenHash,
            'expires_at' => now()->addHours(24),
            'created_at' => now(),
        ]);

        // 5. Send verification email
        try {
            $this->emailService->sendVerificationEmail($user->email, $user->first_name, $rawToken);
        } catch (\Throwable $e) {
            Log::error('Verification email delivery failed', ['userId' => $user->id, 'error' => $e->getMessage()]);
        }

        return ['message' => 'Registration successful. Please check your email to verify your account.'];
    }

    // ─── Verify Email ──────────────────────────────────────────────────────────
    public function verifyEmail(string $rawToken): array
    {
        $tokenHash = CryptoHelper::hashToken($rawToken);

        $record = EmailVerificationToken::with('user')->where('token_hash', $tokenHash)->first();

        if (!$record) {
            throw new \Exception('Invalid verification token', 400);
        }
        if ($record->used) {
            throw new \Exception('This verification link has already been used', 400);
        }
        if ($record->expires_at < now()) {
            throw new \Exception('Verification link has expired. Please request a new one.', 400);
        }
        if ($record->user->status === 'SUSPENDED') {
            throw new \Exception('Account is suspended. Contact support.', 403);
        }

        \DB::transaction(function () use ($record) {
            $record->update(['used' => true]);
            $record->user->update([
                'is_email_verified' => true,
                'status'            => 'PENDING_APPROVAL',
            ]);
        });

        return ['message' => 'Email verified successfully. You can now log in.'];
    }

    // ─── Login ─────────────────────────────────────────────────────────────────
    public function login(array $data, string $ipAddress, string $userAgent): array
    {
        $user = User::where('email', $data['email'])->first();

        // Timing-safe password check
        $passwordValid = false;
        if ($user && $user->password_hash) {
            $passwordValid = Hash::driver('argon2id')->check($data['password'], $user->password_hash);
        } else {
            // Dummy check
            Hash::driver('argon2id')->check($data['password'], Hash::driver('argon2id')->make('dummy'));
        }

        $logAttempt = function (bool $success, ?string $reason = null) use ($user, $data, $ipAddress) {
            try {
                LoginAttempt::create([
                    'id'             => CryptoHelper::generateCuid(),
                    'user_id'        => $user?->id,
                    'email'          => $data['email'],
                    'ip_address'     => $ipAddress,
                    'success'        => $success,
                    'failure_reason' => $reason,
                    'created_at'     => now(),
                ]);
            } catch (\Throwable $e) {}
        };

        if (!$user || !$passwordValid) {
            $logAttempt(false, 'invalid_credentials');
            throw new \Exception('Invalid email or password', 401);
        }

        if (!$user->is_email_verified) {
            $logAttempt(false, 'email_not_verified');
            throw new \Exception('Please verify your email address before logging in', 403);
        }

        if ($user->status === 'SUSPENDED') {
            $logAttempt(false, 'account_suspended');
            throw new \Exception('Your account has been suspended. Contact support.', 403);
        }

        if ($user->status === 'PENDING_APPROVAL') {
            $logAttempt(false, 'account_pending_approval');
            throw new \Exception('Your account is pending admin approval', 403);
        }

        if ($user->status !== 'ACTIVE') {
            $logAttempt(false, 'account_inactive');
            throw new \Exception('Account is not active', 403);
        }

        $logAttempt(true);

        $user->update(['last_login_at' => now()]);

        // 2FA check
        if ($user->is_two_factor_enabled) {
            $tempToken = $this->tokenService->signTempToken($user->id);
            return ['requiresTwoFactor' => true, 'tempToken' => $tempToken];
        }

        return $this->tokenService->issueTokenPair($user->id, $user->email, $user->role, $ipAddress, $userAgent);
    }

    // ─── Refresh Token ─────────────────────────────────────────────────────────
    public function refreshToken(string $rawToken, string $ipAddress, string $userAgent): array
    {
        try {
            $payload = $this->tokenService->verifyRefreshToken($rawToken);
        } catch (\Throwable $e) {
            throw new \Exception('Invalid refresh token. Please log in again.', 401);
        }

        $tokenHash = CryptoHelper::hashToken($rawToken);

        $stored = RefreshToken::with('user')->where('token_hash', $tokenHash)->first();

        if (!$stored) {
            Log::warning('Refresh token not found — possible theft, revoking family', [
                'family' => $payload->family,
                'userId' => $payload->sub,
                'ipAddress' => $ipAddress,
            ]);
            $this->tokenService->revokeTokenFamily($payload->family);
            throw new \Exception('Session invalidated. Possible token theft detected. Please log in again.', 401);
        }

        if ($stored->revoked) {
            Log::warning('Revoked refresh token reused — COMPROMISE SIGNAL, revoking family', [
                'family' => $payload->family,
                'userId' => $stored->user_id,
                'ipAddress' => $ipAddress,
            ]);
            $this->tokenService->revokeTokenFamily($payload->family);
            throw new \Exception('Token reuse detected. All sessions revoked. Please log in again.', 401);
        }

        if ($stored->expires_at < now()) {
            throw new \Exception('Session expired. Please log in again.', 401);
        }

        if ($stored->user->status !== 'ACTIVE') {
            throw new \Exception('Account is not active', 403);
        }

        // Rotate
        $stored->update(['revoked' => true]);

        return $this->tokenService->issueTokenPair(
            $stored->user->id,
            $stored->user->email,
            $stored->user->role,
            $ipAddress,
            $userAgent,
            $payload->family
        );
    }

    // ─── Logout ────────────────────────────────────────────────────────────────
    public function logout(string $rawToken): array
    {
        $tokenHash = CryptoHelper::hashToken($rawToken);
        RefreshToken::where('token_hash', $tokenHash)
            ->where('revoked', false)
            ->update(['revoked' => true]);

        return ['message' => 'Logged out successfully'];
    }

    // ─── Forgot Password ───────────────────────────────────────────────────────
    public function forgotPassword(string $email): array
    {
        $safeMessage = ['message' => 'If that email is registered, you will receive a password reset link shortly.'];

        $user = User::where('email', $email)->first();

        if (!$user || $user->status === 'DELETED') {
            return $safeMessage;
        }

        PasswordResetToken::where('user_id', $user->id)
            ->where('used', false)
            ->update(['used' => true]);

        $rawToken = CryptoHelper::generateSecureToken(32);
        $tokenHash = CryptoHelper::hashToken($rawToken);

        PasswordResetToken::create([
            'id'         => CryptoHelper::generateCuid(),
            'user_id'    => $user->id,
            'token_hash' => $tokenHash,
            'expires_at' => now()->addHour(),
            'created_at' => now(),
        ]);

        try {
            $this->emailService->sendPasswordResetEmail($email, $user->first_name, $rawToken);
        } catch (\Throwable $e) {
            Log::error('Password reset email failed', ['userId' => $user->id, 'error' => $e->getMessage()]);
        }

        return $safeMessage;
    }

    // ─── Reset Password ────────────────────────────────────────────────────────
    public function resetPassword(string $rawToken, string $newPassword): array
    {
        $tokenHash = CryptoHelper::hashToken($rawToken);

        $record = PasswordResetToken::with('user')->where('token_hash', $tokenHash)->first();

        if (!$record) {
            throw new \Exception('Invalid or expired reset token', 400);
        }
        if ($record->used) {
            throw new \Exception('This reset link has already been used', 400);
        }
        if ($record->expires_at < now()) {
            throw new \Exception('Reset link has expired. Please request a new one.', 400);
        }
        if ($record->user->status === 'SUSPENDED') {
            throw new \Exception('Account is suspended. Contact support.', 403);
        }

        $newPasswordHash = Hash::driver('argon2id')->make($newPassword);

        \DB::transaction(function () use ($record, $newPasswordHash) {
            $record->update(['used' => true]);
            $record->user->update(['password_hash' => $newPasswordHash]);
            RefreshToken::where('user_id', $record->user_id)
                ->where('revoked', false)
                ->update(['revoked' => true]);
        });

        try {
            $this->emailService->sendPasswordChangedEmail($record->user->email, $record->user->first_name);
        } catch (\Throwable $e) {
            Log::error('Password changed email failed', ['userId' => $record->user_id]);
        }

        return ['message' => 'Password reset successfully. Please log in with your new password.'];
    }
}

File: Back-ASPNET/src/RecruitmentAuth.Application/Services/AuthService.cs
using BCrypt.Net;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Caching.Memory;
using Microsoft.Extensions.Configuration;
using OtpNet;
using QRCoder;
using Microsoft.Extensions.Logging;
using RecruitmentAuth.Application.DTOs.Auth;
using RecruitmentAuth.Application.Persistence;
using RecruitmentAuth.Domain.Entities;

namespace RecruitmentAuth.Application.Services;

/// <summary>
/// Core authentication use-cases: register, login, 2FA, token refresh, logout,
/// email verification, and password management.
/// </summary>
public class AuthService : IAuthService
{
    // Injected via interface so Application does not depend on Infrastructure types directly.
    private readonly IApplicationDbContext _db;
    private readonly ITokenService _tokenService;
    private readonly IEmailService _emailService;
    private readonly IConfiguration _config;
    private readonly ILogger<AuthService> _logger;
    private readonly IMemoryCache _cache;

    private const int MaxLoginAttempts = 5;
    private const int LockoutMinutes = 15;

    public AuthService(
        IApplicationDbContext db,
        ITokenService tokenService,
        IEmailService emailService,
        IConfiguration config,
        ILogger<AuthService> logger,
        IMemoryCache cache)
    {
        _db = db;
        _tokenService = tokenService;
        _emailService = emailService;
        _config = config;
        _logger = logger;
        _cache = cache;
    }

    // ─── Register ─────────────────────────────────────────────────────────────

    public async Task<AuthResponse> RegisterAsync(RegisterRequest request, string ipAddress, string? userAgent)
    {
        if (await _db.Users.AnyAsync(u => u.Email == request.Email))
            throw new InvalidOperationException("Email already registered.");

        var userId = Guid.NewGuid().ToString();
        var now = DateTime.UtcNow;

        var validRoles = new[] { "CANDIDATE", "RECRUITER" };
        var role = validRoles.Contains(request.Role.ToUpper()) ? request.Role.ToUpper() : "CANDIDATE";

        var user = new User
        {
            Id = userId,
            FirstName = request.FirstName,
            LastName = request.LastName,
            Email = request.Email,
            Phone = request.Phone,
            PasswordHash = BCrypt.Net.BCrypt.HashPassword(request.Password),
            Role = role,
            Status = "PENDING_VERIFICATION",
            IsEmailVerified = false,
            IsTwoFactorEnabled = false,
            CreatedAt = now,
            UpdatedAt = now
        };

        _db.Users.Add(user);

        var rawToken = _tokenService.GenerateRefreshToken();
        var tokenHash = _tokenService.HashToken(rawToken);

        _db.EmailVerificationTokens.Add(new EmailVerificationToken
        {
            Id = Guid.NewGuid().ToString(),
            UserId = userId,
            TokenHash = tokenHash,
            ExpiresAt = now.AddHours(24),
            Used = false,
            CreatedAt = now
        });

        await _db.SaveChangesAsync();

        _ = _emailService.SendVerificationEmailAsync(user.Email, user.FirstName, rawToken)
            .ContinueWith(t => _logger.LogError(t.Exception, "Verification email failed"),
                TaskContinuationOptions.OnlyOnFaulted);

        await LogAuditAsync(userId, "USER_REGISTERED", ipAddress, userAgent);

        return await IssueTokenPairAsync(user, ipAddress, userAgent);
    }

    // ─── Login ────────────────────────────────────────────────────────────────

    public async Task<LoginResult> LoginAsync(LoginRequest request, string ipAddress, string? userAgent)
    {
        var cutoff = DateTime.UtcNow.AddMinutes(-LockoutMinutes);
        var recentFailures = await _db.LoginAttempts
            .CountAsync(a => a.Email == request.Email && a.IpAddress == ipAddress
                          && !a.Success && a.CreatedAt >= cutoff);

        if (recentFailures >= MaxLoginAttempts)
            throw new UnauthorizedAccessException("Too many failed login attempts. Please wait 15 minutes.");

        var user = await _db.Users.FirstOrDefaultAsync(u => u.Email == request.Email);

        bool success = user != null
            && !string.IsNullOrEmpty(user.PasswordHash)
            && BCrypt.Net.BCrypt.Verify(request.Password, user.PasswordHash);

        _db.LoginAttempts.Add(new LoginAttempt
        {
            Id = Guid.NewGuid().ToString(),
            UserId = user?.Id,
            Email = request.Email,
            IpAddress = ipAddress,
            Success = success,
            FailureReason = success ? null : (user == null ? "USER_NOT_FOUND" : "INVALID_PASSWORD"),
            CreatedAt = DateTime.UtcNow
        });
        await _db.SaveChangesAsync();

        if (!success)
            throw new UnauthorizedAccessException("Invalid email or password.");

        if (user!.Status == "PENDING_APPROVAL")
            throw new UnauthorizedAccessException("Your account is pending admin approval.");

        if (user.Status == "SUSPENDED")
            throw new UnauthorizedAccessException("Your account has been suspended.");

        if (user.Status == "DELETED")
            throw new UnauthorizedAccessException("Account not found.");

        user.LastLoginAt = DateTime.UtcNow;
        user.UpdatedAt = DateTime.UtcNow;
        await _db.SaveChangesAsync();

        await LogAuditAsync(user.Id, "USER_LOGIN", ipAddress, userAgent);

        if (user.IsTwoFactorEnabled)
        {
            var tempToken = _tokenService.GenerateTempToken(user.Id);
            return new LoginResult { RequiresTwoFactor = true, TempToken = tempToken };
        }

        var authRes = await IssueTokenPairAsync(user, ipAddress, userAgent);
        return new LoginResult { RequiresTwoFactor = false, AuthResponse = authRes };
    }

    // ─── Two-Factor Authentication ────────────────────────────────────────────

    public async Task<AuthResponse> VerifyTwoFactorAsync(string tempToken, string totpCode,
        string ipAddress, string? userAgent)
    {
        var userId = _tokenService.VerifyTempToken(tempToken);
        if (userId == null)
            throw new UnauthorizedAccessException("Invalid or expired temporary token. Please log in again.");

        var user = await _db.Users.FindAsync(userId);
        if (user == null || !user.IsTwoFactorEnabled || string.IsNullOrEmpty(user.TwoFactorSecret))
            throw new InvalidOperationException("Two-factor authentication is not configured.");

        if (user.Status != "ACTIVE")
            throw new UnauthorizedAccessException("Account is not active.");

        var secretBytes = Base32Encoding.ToBytes(user.TwoFactorSecret);
        var totp = new Totp(secretBytes);
        if (!totp.VerifyTotp(totpCode, out long timeStepMatched, new VerificationWindow(2, 2)))
            throw new UnauthorizedAccessException("Invalid verification code. Please try again.");

        await LogAuditAsync(user.Id, "TWO_FACTOR_LOGIN", ipAddress, userAgent);
        return await IssueTokenPairAsync(user, ipAddress, userAgent);
    }

    public async Task<AuthResponse> VerifyOauthCodeAsync(string tempToken, string code,
        string ipAddress, string? userAgent)
    {
        var userId = _tokenService.VerifyTempToken(tempToken);
        if (userId == null)
            throw new UnauthorizedAccessException("Invalid or expired temporary token. Please log in again.");

        var cacheKey = $"auth:oauth:code:{userId}";
        if (!_cache.TryGetValue(cacheKey, out string? cachedCode))
            throw new UnauthorizedAccessException("Verification session expired. Please log in again.");

        var user = await _db.Users.FindAsync(userId);
        if (user == null)
            throw new InvalidOperationException("User not found.");

        if (code != cachedCode)
            throw new UnauthorizedAccessException("Invalid verification code. Please try again.");

        if (user.Status == "PENDING_APPROVAL")
        {
            _cache.Remove(cacheKey);
            throw new UnauthorizedAccessException("Email verified successfully. Your account is pending activation by an administrator. You will be notified once approved.");
        }

        if (user.Status != "ACTIVE")
            throw new UnauthorizedAccessException("Account is not active.");

        _cache.Remove(cacheKey);

        await LogAuditAsync(user.Id, "OAUTH_LOGIN_VERIFIED", ipAddress, userAgent);
        return await IssueTokenPairAsync(user, ipAddress, userAgent);
    }

    public async Task<object> SetupTwoFactorAsync(string userId)
    {
        var user = await _db.Users.FindAsync(userId)
            ?? throw new KeyNotFoundException("User not found.");

        if (user.IsTwoFactorEnabled)
            throw new InvalidOperationException("Two-factor authentication is already enabled.");

        var secretKey = KeyGeneration.GenerateRandomKey(20);
        var base32Secret = Base32Encoding.ToString(secretKey);

        _cache.Set($"auth:2fa:setup:{userId}", base32Secret, TimeSpan.FromMinutes(10));

        var issuer = "RecruitAuth";
        var account = Uri.EscapeDataString(user.Email);
        var qrCodeUrl = $"otpauth://totp/{issuer}:{account}?secret={base32Secret}&issuer={issuer}&algorithm=SHA1&digits=6&period=30";

        using var qrGenerator = new QRCodeGenerator();
        using var qrCodeData = qrGenerator.CreateQrCode(qrCodeUrl, QRCodeGenerator.ECCLevel.Q);
        using var qrCode = new SvgQRCode(qrCodeData);
        var svgText = qrCode.GetGraphic(20);
        var base64Svg = Convert.ToBase64String(System.Text.Encoding.UTF8.GetBytes(svgText));
        var qrCodeDataUrl = $"data:image/svg+xml;base64,{base64Svg}";

        return new { 
            qrCodeDataUrl = qrCodeDataUrl,
            secret = base32Secret
        };
    }

    public async Task<object> EnableTwoFactorAsync(string userId, string totpCode)
    {
        var cacheKey = $"auth:2fa:setup:{userId}";
        if (!_cache.TryGetValue(cacheKey, out string? pendingSecret))
            throw new UnauthorizedAccessException("Two-factor setup session expired. Please restart setup.");

        var secretBytes = Base32Encoding.ToBytes(pendingSecret);
        var totp = new Totp(secretBytes);
        if (!totp.VerifyTotp(totpCode, out long timeStepMatched, new VerificationWindow(2, 2)))
            throw new UnauthorizedAccessException("Invalid verification code. Please try again.");

        var user = await _db.Users.FindAsync(userId);
        if (user != null)
        {
            user.IsTwoFactorEnabled = true;
            user.TwoFactorSecret = pendingSecret;
            await _db.SaveChangesAsync();
        }

        _cache.Remove(cacheKey);
        await LogAuditAsync(userId, "TWO_FACTOR_ENABLED", null, null);
        return new { message = "Two-factor authentication enabled successfully." };
    }

    public async Task<object> DisableTwoFactorAsync(string userId, string password)
    {
        var user = await _db.Users.FindAsync(userId)
            ?? throw new KeyNotFoundException("User not found.");

        if (!user.IsTwoFactorEnabled)
            throw new InvalidOperationException("Two-factor authentication is not enabled.");

        if (string.IsNullOrEmpty(user.PasswordHash))
            throw new InvalidOperationException("Password-based authentication is not available for this account.");

        if (!BCrypt.Net.BCrypt.Verify(password, user.PasswordHash))
            throw new UnauthorizedAccessException("Incorrect password.");

        user.IsTwoFactorEnabled = false;
        user.TwoFactorSecret = null;
        await _db.SaveChangesAsync();

        await LogAuditAsync(userId, "TWO_FACTOR_DISABLED", null, null);
        return new { message = "Two-factor authentication disabled successfully." };
    }

    // ─── Refresh Token ────────────────────────────────────────────────────────

    public async Task<AuthResponse> RefreshTokenAsync(string refreshToken, string ipAddress, string? userAgent)
    {
        var tokenHash = _tokenService.HashToken(refreshToken);

        var storedToken = await _db.RefreshTokens
            .Include(t => t.User)
            .FirstOrDefaultAsync(t => t.TokenHash == tokenHash);

        if (storedToken == null)
            throw new UnauthorizedAccessException("Invalid refresh token.");

        if (storedToken.Revoked)
        {
            await RevokeTokenFamilyAsync(storedToken.Family);
            throw new UnauthorizedAccessException("Refresh token reuse detected. Please log in again.");
        }

        if (storedToken.ExpiresAt < DateTime.UtcNow)
            throw new UnauthorizedAccessException("Refresh token has expired.");

        storedToken.Revoked = true;
        await _db.SaveChangesAsync();

        await LogAuditAsync(storedToken.UserId, "TOKEN_REFRESHED", ipAddress, userAgent);
        return await IssueTokenPairAsync(storedToken.User, ipAddress, userAgent, storedToken.Family);
    }

    // ─── Logout ───────────────────────────────────────────────────────────────

    public async Task LogoutAsync(string refreshToken)
    {
        var tokenHash = _tokenService.HashToken(refreshToken);
        var storedToken = await _db.RefreshTokens.FirstOrDefaultAsync(t => t.TokenHash == tokenHash);

        if (storedToken != null)
        {
            storedToken.Revoked = true;
            await _db.SaveChangesAsync();
            await LogAuditAsync(storedToken.UserId, "USER_LOGOUT", null, null);
        }
    }

    // ─── Verify Email ─────────────────────────────────────────────────────────

    public async Task<bool> VerifyEmailAsync(string token)
    {
        var tokenHash = _tokenService.HashToken(token);

        var verifyToken = await _db.EmailVerificationTokens
            .Include(t => t.User)
            .FirstOrDefaultAsync(t => t.TokenHash == tokenHash);

        if (verifyToken == null || verifyToken.Used || verifyToken.ExpiresAt < DateTime.UtcNow)
            return false;

        verifyToken.Used = true;
        verifyToken.User.IsEmailVerified = true;
        verifyToken.User.Status = "PENDING_APPROVAL";
        verifyToken.User.UpdatedAt = DateTime.UtcNow;

        await _db.SaveChangesAsync();
        await LogAuditAsync(verifyToken.UserId, "EMAIL_VERIFIED", null, null);
        return true;
    }

    // ─── Forgot / Reset Password ──────────────────────────────────────────────

    public async Task ForgotPasswordAsync(string email, string ipAddress)
    {
        var user = await _db.Users.FirstOrDefaultAsync(u => u.Email == email);
        if (user == null) return;

        var now = DateTime.UtcNow;
        var rawToken = _tokenService.GenerateRefreshToken();
        var tokenHash = _tokenService.HashToken(rawToken);

        _db.PasswordResetTokens.Add(new PasswordResetToken
        {
            Id = Guid.NewGuid().ToString(),
            UserId = user.Id,
            TokenHash = tokenHash,
            ExpiresAt = now.AddHours(1),
            Used = false,
            CreatedAt = now
        });

        await _db.SaveChangesAsync();

        _ = _emailService.SendPasswordResetEmailAsync(user.Email, user.FirstName, rawToken)
            .ContinueWith(t => _logger.LogError(t.Exception, "Password reset email failed"),
                TaskContinuationOptions.OnlyOnFaulted);

        await LogAuditAsync(user.Id, "PASSWORD_RESET_REQUESTED", ipAddress, null);
    }

    public async Task ResetPasswordAsync(string token, string newPassword)
    {
        var tokenHash = _tokenService.HashToken(token);

        var resetToken = await _db.PasswordResetTokens
            .Include(t => t.User)
            .FirstOrDefaultAsync(t => t.TokenHash == tokenHash);

        if (resetToken == null || resetToken.Used || resetToken.ExpiresAt < DateTime.UtcNow)
            throw new InvalidOperationException("Invalid or expired password reset token.");

        resetToken.Used = true;
        resetToken.User.PasswordHash = BCrypt.Net.BCrypt.HashPassword(newPassword);
        resetToken.User.UpdatedAt = DateTime.UtcNow;

        var userTokens = await _db.RefreshTokens
            .Where(t => t.UserId == resetToken.UserId && !t.Revoked)
            .ToListAsync();
        foreach (var t in userTokens) t.Revoked = true;

        await _db.SaveChangesAsync();
        await LogAuditAsync(resetToken.UserId, "PASSWORD_RESET", null, null);
    }

    // ─── Get Current User ─────────────────────────────────────────────────────

    public async Task<UserDto> GetCurrentUserAsync(string userId)
    {
        var user = await _db.Users.FindAsync(userId)
            ?? throw new KeyNotFoundException("User not found.");
        return MapToDto(user);
    }

    // ─── Private Helpers ──────────────────────────────────────────────────────

    private async Task<AuthResponse> IssueTokenPairAsync(
        User user, string ipAddress, string? userAgent, string? existingFamily = null)
    {
        var accessToken = _tokenService.GenerateAccessToken(user);
        var rawRefreshToken = _tokenService.GenerateRefreshToken();
        var refreshHash = _tokenService.HashToken(rawRefreshToken);

        var expiresInStr = _config["JWT_REFRESH_EXPIRES_IN"] ?? "7d";
        var expiresIn = ParseDuration(expiresInStr);

        _db.RefreshTokens.Add(new RefreshToken
        {
            Id = Guid.NewGuid().ToString(),
            UserId = user.Id,
            TokenHash = refreshHash,
            Family = existingFamily ?? Guid.NewGuid().ToString(),
            ExpiresAt = DateTime.UtcNow.Add(expiresIn),
            Revoked = false,
            IpAddress = ipAddress,
            UserAgent = userAgent,
            CreatedAt = DateTime.UtcNow
        });

        await _db.SaveChangesAsync();

        return new AuthResponse
        {
            AccessToken = accessToken,
            RefreshToken = rawRefreshToken,
            User = MapToDto(user)
        };
    }

    private async Task RevokeTokenFamilyAsync(string family)
    {
        var tokens = await _db.RefreshTokens
            .Where(t => t.Family == family && !t.Revoked)
            .ToListAsync();
        foreach (var t in tokens) t.Revoked = true;
        await _db.SaveChangesAsync();
    }

    private async Task LogAuditAsync(string? userId, string eventName, string? ipAddress, string? userAgent)
    {
        _db.AuditLogs.Add(new AuditLog
        {
            Id = Guid.NewGuid().ToString(),
            UserId = userId,
            Event = eventName,
            IpAddress = ipAddress,
            UserAgent = userAgent,
            CreatedAt = DateTime.UtcNow
        });
        await _db.SaveChangesAsync();
    }

    private static UserDto MapToDto(User user) => new()
    {
        Id = user.Id,
        FirstName = user.FirstName,
        LastName = user.LastName,
        Email = user.Email,
        Phone = user.Phone,
        Role = user.Role,
        Status = user.Status,
        IsEmailVerified = user.IsEmailVerified,
        IsTwoFactorEnabled = user.IsTwoFactorEnabled,
        AvatarUrl = user.AvatarUrl,
        CreatedAt = user.CreatedAt,
        LastLoginAt = user.LastLoginAt
    };

    private static TimeSpan ParseDuration(string duration)
    {
        if (duration.EndsWith('d')) return TimeSpan.FromDays(int.Parse(duration[..^1]));
        if (duration.EndsWith('h')) return TimeSpan.FromHours(int.Parse(duration[..^1]));
        if (duration.EndsWith('m')) return TimeSpan.FromMinutes(int.Parse(duration[..^1]));
        return TimeSpan.FromDays(7);
    }
}
